bot/internal/bot/changelog.go

```diff
@@ -0,0 +1,159 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package bot
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"regexp"
+	"strings"
+	"unicode"
+
+	"github.com/gravitational/trace"
+	"k8s.io/utils/strings/slices"
+)
+
+const NoChangelogLabel string = "no-changelog"
+const ChangelogPrefix string = "changelog: "
+const ChangelogTag string = "changelog"
+
+func (b *Bot) Changelog(ctx context.Context) error {
+	pull, err := b.c.GitHub.GetPullRequest(ctx,
+		b.c.Environment.Organization,
+		b.c.Environment.Repository,
+		b.c.Environment.Number,
+	)
+	if err != nil {
+		return trace.Wrap(err, "failed to retrieve pull request for https://github.com/%s/%s/pull/%d", b.c.Environment.Organization, b.c.Environment.Repository, b.c.Environment.Number)
+	}
+
+	if slices.Contains(pull.UnsafeLabels, NoChangelogLabel) {
+		log.Printf("PR contains no changelog label %q, skipping changelog check", NoChangelogLabel)
+		return nil
+	}
+
+	changelogEntry, err := b.getChangelogEntry(ctx, pull.UnsafeBody)
+	if err != nil {
+		return trace.Wrap(err, "failed to get changelog entry")
+	}
+
+	err = b.validateChangelogEntry(ctx, changelogEntry)
+	if err != nil {
+		return trace.Wrap(err, "failed to validate changelog entry")
+	}
+
+	err = b.deleteComments(ctx)
+	if err != nil {
+		return trace.Wrap(err, "failed to delete changelog comments")
+	}
+
+	return nil
+}
+
+func (b *Bot) getChangelogEntry(ctx context.Context, prBody string) (string, error) {
+	changelogRegexString := fmt.Sprintf("(?mi)^%s.*", regexp.QuoteMeta(ChangelogPrefix))
+	changelogRegex, err := regexp.Compile(changelogRegexString)
```

Comment:
I would declare changelog regex as global variable and compile it using regexp.MustCompile. Doesn't really matter from performance perspective since this runs once per job, but would make this function a little cleaner.

Reply:
I can do this but then I have to hard code the entire regex string (`(?mi)^changelog: .*` in this case) as I cannot call functions when defining a const. The downside of this is that it would be _slightly_ harder to update the changelog regex in the future, and makes the pattern matching logic a little harder to read (IMO). Would you like me to continue moving the regex string to a const?